src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -249,11 +249,14 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
+            var pair = new KeyValuePair<Type, MethodInfo>();
+
             foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
                 if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    return xEquatableArgument.Value;
+                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
```

Comment:
@rprouse @CharliePoole Sorry, I didn't get back to you soon; I have been a bit busy.

The idea is as follows:
If there exists any interface with the type of `second`, we take that. If any other interface also matches, we replace the previous interface, if the other is inheriting from the current chosen one; otherwise we ignore it.

This ensures that we always find one, if there is one. It does, however, not ensure that we get the best interface. Consider:

```
interface IFoo{}
interface IBar : IFoo;
interface IBaz : IFoo;
interface IQux : IBaz;
```
We will always end up with either `IEquatable<IBar>` or `IEquatable<IQux>` depending on which interface we see first:
- if we first see `IEquatable<IFoo>`, it depends on the order of the next interface (see two options below)
- if we first see `IEquatable<IBar>`, we end up with that.
- if we first see `IEquatable<IBaz>` or `IEquatable<IQux>`, we end up with `IEquatable<IQux>`.

I don't think there is a best choice here, so I'm not sure if there is a best way to handle this. If the first object implements `IEquatable` for different unrelated types of `second` then there might be a problem that we cannot easily solve.